### PR TITLE
CompiledPageRouteModelProvider should de-dup descriptors

### LINKED ISSUE #7543: Page descriptors duplicated
I'm using precompilation for UI in a library and at least one page descriptor is getting duplicated.
@pranavkm @rynowak 

image

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #7543

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs [hunk 1]
```diff
@@ -56,22 +56,55 @@ public void OnProvidersExecuted(PageRouteModelProviderContext context)
             }
         }
 
-        /// <summary>
-        /// Gets the sequence of <see cref="CompiledViewDescriptor"/> from <paramref name="applicationManager"/>.
-        /// </summary>
-        /// <param name="applicationManager">The <see cref="ApplicationPartManager"/>s</param>
-        /// <returns>The sequence of <see cref="CompiledViewDescriptor"/>.</returns>
-        protected virtual IEnumerable<CompiledViewDescriptor> GetViewDescriptors(ApplicationPartManager applicationManager)
+        private IEnumerable<CompiledViewDescriptor> GetViewDescriptors(ApplicationPartManager applicationManager)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: Why are there duplicates? What's the root cause?

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: The feature provider does not de-dup across multiple application parts. For instance when the app overrides a class library page. There's also the possibility that different feature providers have duplicates (we had a test in precompilation that ran the `MvcRazorCompile` target on a project that built using the Razor Sdk. 

We do the same thing in the RazorViewCompiler where we pick the first view for a given path. 

Going back to your question, in @javiercn's case, the issue was the view was stamped with the default compiled view factory (not the null one) and `ConfigureB2C` was re-adding the part.

Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: OK

Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: @DamianEdwards wanted a feature here where we could throw on conflicts when we don't have a clear way to disambiguate.

The principle is that the application can override anything, but conflicts between libraries are genuine conflicts.

I'm not sure that we're in the position to add something like this late in the game. It feels like it needs to be a first class feature to me.

Comment (javiercn) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: We would need to keep information about which part the view came from and information about the "rank of the parts".

This second part we can just collect when we are doing the initial ordering and save it in something like a "RankFeatureProvider" that can be used to compute the Rank of each part and make decisions on.

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: Yup. We don't have sufficient information about the order of parts by the time we get to here. Like @javiercn said, we'll need some way to stash this metadata as part of assembly \ part discovery that gets used later.

**Pull request closed without merging**